klog.go

```diff
@@ -824,21 +831,50 @@ func kvListFormat(b *bytes.Buffer, keysAndValues ...interface{}) {
 		}
 		b.WriteByte(' ')
 
-		switch v.(type) {
-		case string, error:
-			b.WriteString(fmt.Sprintf("%s=%q", k, v))
+		switch v := v.(type) {
+		case string:
+			writeStringValue(b, k, true, v)
+		case error:
+			writeStringValue(b, k, true, v.Error())
 		case []byte:
 			b.WriteString(fmt.Sprintf("%s=%+q", k, v))
 		default:
-			if _, ok := v.(fmt.Stringer); ok {
-				b.WriteString(fmt.Sprintf("%s=%q", k, v))
+			if s, ok := v.(fmt.Stringer); ok {
+				writeStringValue(b, k, true, s.String())
 			} else {
-				b.WriteString(fmt.Sprintf("%s=%+v", k, v))
+				writeStringValue(b, k, false, fmt.Sprintf("%+v", v))
 			}
 		}
 	}
 }
 
+func writeStringValue(b *bytes.Buffer, k interface{}, quoteV bool, v string) {
+	if !strings.Contains(v, "\n") {
+		b.WriteString(fmt.Sprintf("%s=", k))
+		if quoteV {
+			// Simple string, quote quotation marks and non-printable characters.
+			b.WriteString(strconv.Quote(v))
+			return
+		}
+		// Non-string with no line breaks.
+		b.WriteString(v)
+		return
+	}
+	// Complex multi-line string, show as-is with indention.
+	b.WriteString(fmt.Sprintf("===start of %s===\n ", k))
+	writeString(b, v)
+	b.WriteString(fmt.Sprintf("\n ===end of %s===", k))
+}
+
+func writeString(b *bytes.Buffer, s string) {
+	for i, line := range strings.Split(s, "\n") {
```

Comment:
Does strings.Split here needlessly copies the string into slices?

Reply:
Yes. I'll add benchmarking and make the code more efficient before merging. Right now I am only looking for feedback on the format.

/hold

Reply:
https://github.com/kubernetes/klog/pull/273 has detailed benchmark data.